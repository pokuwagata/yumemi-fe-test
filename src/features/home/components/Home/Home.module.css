.layout {
  max-width: 960px;
  min-height: 100dvh;
  padding: var(--space2) var(--space4);
  margin: 0 auto;
  background-color: var(--stale-50);
}

.header {
  margin-bottom: var(--space4);
}

.prefectures {
  margin-bottom: var(--space4);
}
