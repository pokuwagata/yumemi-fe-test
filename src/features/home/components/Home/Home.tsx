import styles from "./Home.module.css";

import { Chart } from "~/features/home/components/Chart";
import { Header } from "~/features/home/components/Header/Header";
import { Prefectures } from "~/features/home/components/Prefectures";
import { SelectedPrefCodesContextProvider } from "~/features/home/contexts/SelectedPrefCodesContext";
import { Prefecture } from "~/types/api";

type Props = {
  prefectures: Prefecture[];
};

export function Home({ prefectures }: Props) {
  return (
    <div className={styles.layout}>
      <div className={styles.header}>
        <Header>ゆめみフロントエンドコーディング試験</Header>
      </div>
      <main>
        <section>
          <SelectedPrefCodesContextProvider>
            <div className={styles.prefectures}>
              <Prefectures prefectures={prefectures} />
            </div>
            <Chart prefectures={prefectures} />
          </SelectedPrefCodesContextProvider>
        </section>
      </main>
    </div>
  );
}
