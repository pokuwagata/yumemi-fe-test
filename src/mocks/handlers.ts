import { http, HttpResponse } from "msw";

import population1 from "./population-1.json";
import population2 from "./population-2.json";
import prefectures from "./prefectures.json";

import { API_BASE_URL, PROXY_API_PATH } from "~/lib/const";

// @see https://opendata.resas-portal.go.jp/docs/api/v1/detail/index.html
export function getErrorDetail(code: string) {
  let statusCode;
  let body;

  switch (code) {
    case "400":
      statusCode = 200;
      body = "400";

      break;
    case "403":
      statusCode = 200;

      body = {
        statusCode: "403",
        message: "Forbidden.",
        description: "",
      };

      break;
    case "404-1":
      statusCode = 200;

      body = {
        statusCode: "404",
        message: "404. That's an error.",
        description: "The requested URL /404 was not found on this server.",
      };

      break;
    case "404-2":
      statusCode = 200;
      body = "404";
      break;
    case "429":
      statusCode = 429;

      body = {
        message: null,
      };

      break;
    case "500":
      statusCode = 500;

      body = "Server Error";
      break;
  }

  return { statusCode, body };
}

export const handlers = [
  http.get(`${API_BASE_URL}/v1/prefectures`, () => {
    return HttpResponse.json(prefectures);
  }),
  http.get(`${PROXY_API_PATH}/population`, ({ request, cookies }) => {
    const errorCode = cookies.error;

    if (errorCode) {
      // Cookie にエラーコードが設定されている場合はエラーレスポンスを返す
      const { statusCode, body } = getErrorDetail(errorCode);

      return HttpResponse.json(body, {
        status: statusCode,
      });
    }

    const url = new URL(request.url);

    const prefCode = url.searchParams.get("prefCode");

    let population;

    switch (prefCode) {
      case "1":
        population = population1;
        break;
      case "2":
        population = population2;
        break;
      default:
        population = population1;
    }

    return HttpResponse.json(population);
  }),
];
