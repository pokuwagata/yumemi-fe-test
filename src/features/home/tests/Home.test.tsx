import "@testing-library/jest-dom/vitest";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { ReactElement } from "react";
import { beforeAll, afterEach, afterAll } from "vitest";

import { Home } from "~/features/home/components/Home";
import prefectures from "~/mocks/prefectures.json";
import { server } from "~/mocks/server";

beforeAll(() => {
  global.ResizeObserver = class ResizeObserver {
    observe() {}
    unobserve() {}
    disconnect() {}
  };

  server.listen();
});

afterEach(() => server.resetHandlers());
afterAll(() => server.close());

// @see https://github.com/recharts/recharts/issues/2982#issuecomment-1545948222
vi.mock("recharts", async () => {
  const OriginalModule =
    await vi.importActual<typeof import("recharts")>("recharts");

  return {
    ...OriginalModule,
    ResponsiveContainer: ({ children }: { children: ReactElement }) => (
      <OriginalModule.ResponsiveContainer width={800} height={800}>
        {children}
      </OriginalModule.ResponsiveContainer>
    ),
  };
});

describe("正常系", () => {
  test("初期表示時に北海道が選択されてグラフが表示されていること", async () => {
    const { container } = render(<Home prefectures={prefectures.result} />);

    const checkbox = await screen.findByLabelText("北海道");
    expect(checkbox).toBeChecked();

    const radio = screen.getByLabelText("総人口");
    expect(radio).toBeChecked();

    await waitFor(() => {
      const line = container.querySelector("path[name='北海道']");

      expect(line).toBeInTheDocument();
    });
  });

  test("複数の都道府県を選択しグラフが表示されていること", async () => {
    const { container } = render(<Home prefectures={prefectures.result} />);

    const checkbox = await screen.findByLabelText("青森県");

    const user = userEvent.setup();
    await user.click(checkbox);

    await waitFor(() => {
      const line1 = container.querySelector("path[name='北海道']");
      const line2 = container.querySelector("path[name='青森県']");

      expect(line1).toBeInTheDocument();
      expect(line2).toBeInTheDocument();
    });
  });

  test("人口動態の種別を切り替えると対応するグラフが表示されていること", async () => {
    const { container } = render(<Home prefectures={prefectures.result} />);

    const checkbox = await screen.findByLabelText("北海道");

    expect(checkbox).toBeChecked();

    const radio = screen.getByLabelText("年少人口");

    const user = userEvent.setup();
    await user.click(radio);
    expect(radio).toBeChecked();

    await waitFor(() => {
      const line = container.querySelector("path[name='北海道']");

      expect(line).toMatchSnapshot();
    });
  });

  test("都道府県が未選択の場合は人口動態の種別を切り替えられないこと", async () => {});
  test("都道府県が未選択の場合はグラフが非表示になっていること", async () => {});
});
