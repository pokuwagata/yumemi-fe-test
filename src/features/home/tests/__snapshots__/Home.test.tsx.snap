// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`正常系 > 人口動態の種別を切り替えると対応するグラフが表示されていること 1`] = `
<path
  class="recharts-curve recharts-line-curve"
  d="M85,136.8966033333333C98.62745098039215,129.33688555555554,112.25490196078431,121.77716777777779,125.88235294117646,120.76433333333334C139.50980392156862,119.75149888888889,153.13725490196077,120.25791611111111,166.76470588235293,119.24508166666666C180.39215686274508,118.23224722222221,194.01960784313727,108.46111611111114,207.64705882352942,100.51827000000003C221.27450980392157,92.57542388888892,234.90196078431373,78.50745194444447,248.52941176470588,71.58800500000001C262.156862745098,64.66855805555556,275.78431372549016,59.001588333333316,289.4117647058823,59.001588333333316C303.0392156862745,59.001588333333316,316.6666666666667,63.35628166666668,330.29411764705884,63.35628166666668C343.921568627451,63.35628166666668,357.54901960784315,57.43427833333331,371.1764705882353,57.43427833333331C384.80392156862746,57.43427833333331,398.4313725490196,57.809782222222196,412.05882352941177,58.56078999999997C425.6862745098039,59.31179777777774,439.3137254901961,61.7100708333333,452.94117647058823,65.29199833333331C466.5686274509804,68.87392583333332,480.19607843137254,75.06394055555559,493.8235294117647,80.05235500000003C507.45098039215685,85.04076944444448,521.0784313725491,89.50810583333333,534.7058823529412,95.22248499999999C548.3333333333334,100.93686416666665,561.9607843137255,106.93361138888886,575.5882352941177,114.33862999999998C589.2156862745098,121.7436486111111,602.843137254902,130.87187277777778,616.4705882352941,139.65259666666668C630.0980392156863,148.43332055555558,643.7254901960785,157.48842305555553,657.3529411764706,167.0229733333333C670.9803921568628,176.5575236111111,684.6078431372549,186.49460805555555,698.2352941176471,196.85989833333332C711.8627450980392,207.2251886111111,725.4901960784314,218.23665055555557,739.1176470588235,229.214715C752.7450980392157,240.19277944444445,766.3725490196078,251.4605322222222,780,262.72828499999997"
  fill="none"
  height="730"
  name="北海道"
  stroke="#e7417b"
  stroke-width="1"
  width="695"
/>
`;
